bpl/neutral_dixon_coles.py

```diff
@@ -0,0 +1,492 @@
+            defence_prior_mean = mean_defence + jnp.matmul(
+                standardised_covariates, defence_coefficients[..., None]
+            ).squeeze(-1)
+        else:
+            attack_prior_mean = 0.0
+            defence_prior_mean = mean_defence
+
+        with numpyro.plate("teams", num_teams):
+            standardised_attack = numpyro.sample(
+                "standardised_attack", dist.Normal(loc=0.0, scale=1.0)
+            )
+            standardised_defence = numpyro.sample(
+                "standardised_defence",
+                dist.Normal(
+                    loc=rho * standardised_attack, scale=jnp.sqrt(1.0 - rho**2.0)
+                ),
+            )
+
+            with reparam(config={"home_attack_advantage": LocScaleReparam(centered=0)}):
+                home_attack_advantage = numpyro.sample(
+                    "home_attack_advantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
+                )
+            with reparam(
+                config={"away_attack_disadvantage": LocScaleReparam(centered=0)}
+            ):
+                away_attack_disadvantage = numpyro.sample(
+                    "away_attack_disadvantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
+                )
+            with reparam(
+                config={"home_defence_advantage": LocScaleReparam(centered=0)}
+            ):
+                home_defence_advantage = numpyro.sample(
+                    "home_defence_advantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
+                )
+            with reparam(
+                config={"away_defence_disadvantage": LocScaleReparam(centered=0)}
+            ):
+                away_defence_disadvantage = numpyro.sample(
+                    "away_defence_disadvantage",
+                    dist.Normal(mean_home_advantage, std_home_advantage),
+                )
+        attack = numpyro.deterministic(
+            "attack", attack_prior_mean + standardised_attack * std_attack
+        )
+        defence = numpyro.deterministic(
+            "defence", defence_prior_mean + standardised_defence * std_defence
+        )
+
+        expected_home_goals = jnp.exp(
+            attack[home_team]
+            - defence[away_team]
+            + (1 - neutral_venue) * home_attack_advantage[home_team]
+            + (1 - neutral_venue) * away_defence_disadvantage[away_team]
+        )
+        expected_away_goals = jnp.exp(
+            attack[away_team]
+            - defence[home_team]
+            - (1 - neutral_venue) * away_attack_disadvantage[away_team]
+            - (1 - neutral_venue) * home_defence_advantage[home_team]
+        )
+
+        # FIXME: this is because the priors allow crazy simulated data before inference
+        expected_home_goals = jnp.clip(expected_home_goals, a_max=15.0)
+        expected_away_goals = jnp.clip(expected_away_goals, a_max=15.0)
```

Comment:
not 100% sure if this is the right fix - but I think we need to bound the correlation term (like it does in the paper). The following code just places a uniform distribution between the minimum and maximum defined in the paper
```
# impose bounds on the correlation coefficient
corr_coef_raw = numpyro.sample("corr_coef_raw", dist.Uniform(low=0.0, high=1.0))
UB = jnp.min(jnp.array([jnp.min(1.0/(expected_home_goals*expected_away_goals)), 1]))
LB = jnp.max(jnp.array([jnp.max(-1.0/expected_home_goals), jnp.max(-1.0/expected_away_goals)]))
corr_coef = numpyro.deterministic("corr_coef", LB + corr_coef_raw * (UB-LB))
```
and this would mean deleting the original definition of `corr_coef` above

Reply:
Not sure about this one, need to dig into it/have a chat at some point to understand. All of this is carried over from the 'normal' BPL model, it's not been changed for the neutral model.

Reply:
I'm happy to ignore this for now - just a thought that could be why you had these as a quick fix. The paper has bounds for this and so thought would be good to implement them too

Reply:
I've had a look back at the paper and see what you're doing here now. Wondering whether a truncated normal would be better than uniform (it would be closer to what we're doing now at least)? We'd also need to update this in all the models, not just the Neutral one (at least if changing dixon_coles_correlation_term).

Reply:
Truncated normal may have too low probability in the tails possibly, maybe something inbetwteen the two

Reply:
(e.g. truncated normal with a larger variance)

Reply:
We can maybe do a beta(2,2) distribution? It's bounded between 0 and 1 so it can work nicely I think if we wanted something bit more informative than a uniform

Reply:
I was wondering about beta but I think rho can be negative looking at the paper (also in what you did above)

Reply:
Yeah that's what I was doing (should've commented the code better!):

I sample from an initial distribution which is bounded (so this is either uniform or beta), the `_raw` variable, and then transform this using the bounds which gives me a distribution in `[LB, UB]` where they are defined in the paper and I take the minimum or maximum of these values computed using `expected_home_goals` and `expected_away_goals` (which refer to `lambda` and `mu` in the paper respectively).